### PR TITLE
Collect and Store etcd Snapshots and Metadata

### PR DESCRIPTION
**Pull request by briandowns**
#### Proposed Changes ####

This PR adds the ability to collect and store the active snapshots on the node with their name, creation date, and size. It also adds the node name to the name of the snapshot so snapshots stored in S3 aren't overwritten as each node writes it's snapshot to the bucket. The retention logic has been updated to account for the snapshot name prefix update.

Docs should be updated to reflect these changes and additions.

#### Linked Issues ####

#3234

### CODE REVIEW

File: pkg/etcd/etcd.go
Comment (brandond) [inline pkg/etcd/etcd.go:97]: Since you're not using the second parameter, you could probably just move this into a util package and consolidate: https://github.com/k3s-io/k3s/blob/8f1a20c0d32c957c8ef953ad44bac8f9f53220e0/pkg/agent/run.go#L136

File: pkg/etcd/etcd.go
Comment (brandond) [inline pkg/etcd/etcd.go:983]:
```suggestion
					Name:      version.Program + "-etcd-snapshots",
```
but also make it a variable since it's repeated down below in the error handlers

**Review by brandond — CHANGES_REQUESTED** [review]
the configmap create/update code seems really gnarly, is there some library code we can use to make that a little more compact?

**Comment by briandowns** [comment]
> The configmap create/update code seems really gnarly, is there some library code we can use to make that a little more compact?
> 
> Also, there seem to be a bunch of new empty files checked in to vendor - guessing `go mod vendor` hosed something up again.

Yup, 

> The configmap create/update code seems really gnarly, is there some library code we can use to make that a little more compact?
> 
> Also, there seem to be a bunch of new empty files checked in to vendor - guessing `go mod vendor` hosed something up again.

Yup, focused on logic flow. After that's verified, performing clean up and organization.

File: pkg/etcd/etcd.go
Comment (brandond) [inline pkg/etcd/etcd.go:906]: Kubernetes has a metav1.Time that is handles serializing/deserializing timestamps to/from json and protobuf:
https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#Time

Comment (briandowns) [reply pkg/etcd/etcd.go:906]: Thanks.

File: pkg/etcd/etcd.go
Comment (dweomer) [inline pkg/etcd/etcd.go:958]:
```suggestion
func updateSnapshotData(data map[string]string, snapshotFiles []snapshotFile) error {
```

less confusing name

File: pkg/etcd/etcd.go
Comment (dweomer) [inline pkg/etcd/etcd.go:983]: This shouldn't be necessary, the controller should fill this in for you when you ship it to the apiserver

File: pkg/cluster/etcd.go
Comment (dweomer) [inline pkg/cluster/etcd.go:9]: We are attempting to establish a k8s client connection in a package-level init function?

Comment (briandowns) [reply pkg/cluster/etcd.go:9]: Since remove between them and last push.

**Review by brandond — APPROVED** [review]
Looks great!

**Comment by cjellick** [comment]
@briandowns - Post an example of what the contents of the config map will look like here and then get @ibuildthecloud's eyes on it. Let's ensure that the data structure meets his needs.

Once you post it, ping me too so I can bug darren to look at it.

**Comment by briandowns** [comment]
The ConfigMap data structure is a `map[string]string` with the key of the map being the name of the snapshot itself. The value is a JSON serialized string with the structure below. It differs only slightly from what I had discussed with @ibuildthecloud earlier today where I needed to change the key from the full path ot just the name and opted for a "location" field to contain the full path and have the idea be uniform across local and S3 storage locations.

```go
type snapshotFile struct {
	Name string `json:"name"`
	// Location contains the full path of the snapshot. For
	// local paths, the location will be prefixed with "file://"
	// and for S3 will be "s3://".
	Location  string       `json:"location"`
	NodeName  string       `json:"nodeName"`
	CreatedAt *metav1.Time `json:"createdAt"`
	Size      int64        `json:"size"`
}
```

ConfigMap with snapshots stored locally.

```json
{
    "apiVersion": "v1",
    "data": {
        "etcd-snapshot-etcd-testing-01-1619652180": "{\"name\":\"etcd-snapshot-etcd-testing-01-1619652180\",\"location\":\"s3://k3s-snapshots/etcd-snapshot-etcd-testing-01-1619652180\",\"nodeName\":\"\",\"createdAt\":\"2021-04-28T23:23:02Z\",\"size\":2940960}"
    },
    "kind": "ConfigMap",
    "metadata": {
        "creationTimestamp": "2021-04-28T23:07:00Z",
        "name": "k3s-etcd-snapshots",
        "namespace": "kube-system",
        "resourceVersion": "3383",
        "uid": "e4322241-3ff8-4b32-9e38-483b13e1f623"
    }
}
```

ConfigMap with snapshots stored in S3.

```json
{
    "apiVersion": "v1",
    "data": {
        "etcd-snapshot-etcd-testing-01-1619653860": "{\"name\":\"etcd-snapshot-etcd-testing-01-1619653860\",\"location\":\"s3://k3s-snapshots/etcd-snapshot-etcd-testing-01-1619653860\",\"nodeName\":\"\",\"createdAt\":\"2021-04-28T23:51:01Z\",\"size\":3854368}"
    },
    "kind": "ConfigMap",
    "metadata": {
        "creationTimestamp": "2021-04-28T23:07:00Z",
        "name": "k3s-etcd-snapshots",
        "namespace": "kube-system",
        "resourceVersion": "3545",
        "uid": "e4322241-3ff8-4b32-9e38-483b13e1f623"
    }
}
```

**Comment by brandond** [comment]
Hmm, why is the node name coming out empty?

**Comment by briandowns** [comment]
> Hmm, why is the node name coming out empty?

Was a bug I fixed but copied the output. I'll edit it.

**Comment by brandond** [comment]
Since the name is both the key and in the struct itself, is it even worth using a map, or would a list be simpler? No point in having the same data in there twice.

**Comment by cjellick** [comment]
I had the same thought, but let it go. I don't feel strongly. Agree with your offline comment: whatever is easiest for @ibuildthecloud to consume.

**Review by dweomer — APPROVED** [review]

**Comment by ibuildthecloud** [comment]
@briandowns I need this format to be as close to possible to the arguments that I need to pass back to `k3s server.`  Specifically bucket, folder, region for s3.  I'd rather not have to parse that data out of Location.  Also it would be nice if Name corresponded to the value that will be passed to `--cluster-reset-restore-path`.

File: pkg/etcd/etcd.go
Comment (brandond) [inline pkg/etcd/etcd.go:873]: @ibuildthecloud are these somewhere that we could reference in rancher/rancher, or do we just need to maintain two copies of the struct definition?

File: pkg/etcd/etcd.go
Comment (brandond) [inline pkg/etcd/etcd.go:921]: For local files this is just the basename, not the full path including the folder, which is what I believe the Key is for S3 objects?

```suggestion
				Name:     filepath.Base(obj.Key),
```

File: pkg/etcd/etcd.go
Comment (brandond) [inline pkg/etcd/etcd.go:987]: I think the pattern elsewhere in the codebase is to just sleep for a couple seconds?

File: pkg/etcd/etcd.go
Comment (brandond) [inline pkg/etcd/etcd.go:990]: Probably worth assigning `e.config.Runtime.Core.Core().V1().ConfigMap()` to a variable once it's been confirmed that Runtime.Core is non-nil?

File: pkg/etcd/etcd.go
Comment (brandond) [inline pkg/etcd/etcd.go:889]: It sounds like we don't need a generic Location, since:
* For S3 restores we get the bucket/folder info from the S3 field
* For local restores we just need the filename and directory (which could perhaps go in a separate field?)

Comment (briandowns) [reply pkg/etcd/etcd.go:889]: This needs a comment update.

File: pkg/etcd/etcd.go
Comment (briandowns) [reply pkg/etcd/etcd.go:990]: More a matter of style in that regard over function. I don't have a strong opinion either way.

File: pkg/etcd/etcd.go
Comment (briandowns) [reply pkg/etcd/etcd.go:987]: This is a more efficient approach that dequeues the goroutine in favor of another during the startup routine where a sleep wouldn't. I did this so not to hinder the start up speed when running with very short snapshot crons.

**Review by brandond — APPROVED** [review]

**Comment by cjellick** [comment]
Going to capture here some of the issues we identified that need addressed:
1. HA setup. Node A writes to config map. The Node B does. Node B wipes out Node A’s entries in the config map.
2. After restore: when you do a restore, the configmap reverts to an older version of itself. It will be inaccurate until the next snapshotting runs (which would presumably be an infrequent cron).

File: pkg/etcd/etcd.go
Comment (brandond) [inline pkg/etcd/etcd.go:902]: How does this work on a bucket with many objects? @ibuildthecloud mentioned that folks frequently dump backups into a bucket with a bunch of other stuff; there could potentially be thousands of objects here.

Comment (briandowns) [reply pkg/etcd/etcd.go:902]: Given the time frame I haven't had the opportunity to test at that scale. In my own bucket, I've tested with a few hundred and there's nothing different than just a couple.

File: pkg/etcd/etcd.go
Comment (cjellick) [inline pkg/etcd/etcd.go:903]: Going to remove before merging?

File: pkg/cli/etcdsnapshot/etcd_snapshot.go
Comment (briandowns) [reply pkg/cli/etcdsnapshot/etcd_snapshot.go:43]: I completely agree. However, I was following the pattern set in the repo in the agent code and what was used previously in the etcd code. Should I still add the field?

Comment (brandond) [reply pkg/cli/etcdsnapshot/etcd_snapshot.go:43]: The convention in the K3s codebase is to pass this between controllers via the NODE_NAME environment variable so I think it's fine.

File: pkg/cluster/cluster.go
Comment (cjellick) [inline pkg/cluster/cluster.go:113]: Might be fine, just curious: why put this in a goroutine?

File: pkg/cluster/cluster.go
Comment (cjellick) [inline pkg/cluster/cluster.go:115]: We are going to just log and forget this error, so I think it should be wrapped with a message providing context and making it unique. Something like (but not necessarily exactly) "Failed to record snapshots for cluster."

**Review by cjellick — APPROVED** [review]
lgtm, assuming it goes green

**Pull request merged**
